File: android/src/main/kotlin/com/artahc/rx_method_channel/RxMethodChannel.kt
package com.artahc.rx_method_channel

import android.util.Log
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.util.logging.Level
import java.util.logging.Logger

typealias SingleContainer = (Argument) -> Single<Any>
typealias CompletableContainer = (Argument) -> Completable
typealias ObservableContainer = (Argument) -> Observable<Any>

enum class Action(val value: String) {

    Cancel("cancel"), Subscribe("subscribe");

    companion object {
        private val map = Action.values().associateBy(Action::value)
        operator fun get(value: String) = map[value]
    }
}

enum class MethodType(val value: String) {
    Single("single"), Completable("completable"), Observable("observable");

    companion object {
        private val map = MethodType.values().associateBy(MethodType::value)
        operator fun get(value: String) = map[value]
    }
}

class RxMethodChannel(channelName: String, binaryMessenger: BinaryMessenger) :
    MethodChannel.MethodCallHandler {
    private val tag = "RxMethodChannelPlugin"
    private val logger = Logger.getLogger(tag)
    private val channel: MethodChannel

    init {
        channel = MethodChannel(binaryMessenger, channelName)
        channel.setMethodCallHandler(this)
    }

    private val registeredSingle = mutableMapOf<String, SingleContainer>()
    private val registeredObservable = mutableMapOf<String, ObservableContainer>()
    private val registeredCompletable = mutableMapOf<String, CompletableContainer>()

    private val subscriptions = mutableMapOf<Int, Disposable>()

    fun <T> registerSingle(methodName: String, call: SingleContainer) {
        logger.log(Level.INFO, "Registered single: $methodName")
        registeredSingle[methodName] = call
    }

    fun registerCompletable(methodName: String, call: CompletableContainer) {
        logger.log(Level.INFO, "Registered completable $methodName")
        registeredCompletable[methodName] = call
    }

    fun <T> registerObservable(methodName: String, call: ObservableContainer) {
        logger.log(Level.INFO, "Registered observable $methodName")
        registeredObservable[methodName] = call
    }

    private fun removeSubscription(requestId: Int) {
        logger.log(
            Level.INFO,
            "Disposed subscription with requestId: $requestId -> ${subscriptions[requestId]}"
        )
        subscriptions[requestId]?.dispose()
        subscriptions.remove(requestId)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        logger.log(Level.INFO, "Invoking ${call.method} ${call.arguments}")

        val args = call.arguments as HashMap<String, Any>
        val requestId = args["requestId"] as Int

        when (Action[call.method]) {
            Action.Cancel -> {
                removeSubscription(requestId)
                result.success(null)
            }
            Action.Subscribe -> {
                val methodName = args["methodName"] as String
                val methodArgument = args["arguments"] as HashMap<String, Any>
                val rawMethodType = args["methodType"] as String

                when (MethodType[rawMethodType]) {
                    MethodType.Single -> {
                        if (!registeredSingle.containsKey(methodName)) {
                            result.error(
                                METHOD_NOT_FOUND,
                                "Method $methodName is not registered.",
                                null
                            )
                            return
                        }

                        val source = registeredSingle[methodName]
                        subscriptions[requestId] = source!!.invoke(methodArgument)
                            .doOnTerminate {
                                removeSubscription(requestId)
                            }
                            .subscribe({
                                result.success(it)
                            }, { error ->
                                result.error(OPERATION_ERROR, error.localizedMessage, error)
                            })
                    }
                    MethodType.Completable -> {
                        if (!registeredCompletable.containsKey(methodName)) {
                            result.error(
                                METHOD_NOT_FOUND,
                                "Method $methodName is not registered.",
                                null
                            )
                            return
                        }

                        val source = registeredCompletable[methodName]
                        subscriptions[requestId] = source!!.invoke(methodArgument)
                            .doOnTerminate {
                                removeSubscription(requestId)
                            }
                            .subscribe({
                                result.success(null)
                            }, { error ->
                                result.error(OPERATION_ERROR, error.localizedMessage, error)
                            })
                    }
                    MethodType.Observable -> {
                        if (!registeredObservable.containsKey(methodName)) {
                            result.error(
                                METHOD_NOT_FOUND,
                                "Method $methodName is not registered.",
                                null
                            )
                            return
                        }

                        val source = registeredObservable[methodName]
                        subscriptions[requestId] = source!!.invoke(methodArgument)
                            .observeOn(AndroidSchedulers.mainThread())
                            .doOnComplete {
                                result.success(null)
                                removeSubscription(requestId)
                            }
                            .subscribe({
                                channel.invokeMethod(
                                    "observableCallback",
                                    ObservableCallback(requestId, it).toJson()
                                )
                            }, { error ->
                                result.error(OPERATION_ERROR, error.localizedMessage, null)
                                removeSubscription(requestId)
                            })
                    }
                    null -> result.error(
                        INVALID_OPERATION,
                        "Invalid method type: $rawMethodType",
                        null
                    )
                }
            }
            null -> result.error(INVALID_OPERATION, "Invalid operation: ${call.method}", null)
        }
    }
}

File: macos/Classes/Internal/RxMethodChannel.swift
//
//  RxMethodChannel.swift
//  rx_method_channel
//
//  Created by Komang Arta Wibawa on 31/01/23.
//

import Foundation
import RxSwift
import FlutterMacOS

public typealias Argument = [String: Any]
public typealias SingleSource = (Argument) -> Single<Any>
public typealias CompletableSource = (Argument) -> Completable
public typealias ObservableSource = (Argument) -> Observable<Any>

let METHOD_NOT_FOUND = "1"
let OPERATION_ERROR = "2"
let INVALID_OPERATION = "3"

enum OperationType: String{
    case cancel
    case subscribe
}

enum MethodType: String {
    case single
    case completable
    case observable
}

public class RxMethodChannel {
    private let channel: FlutterMethodChannel
    private let channelName: String
    
    private var registeredSingle : [String: SingleSource] = [:]
    private var registeredCompletable : [String: CompletableSource] = [:]
    private var registeredObservable: [String: ObservableSource] = [:]
    
    private var subscriptions: [Int:Disposable] = [:]
    
    public init(channelName: String, binaryMessenger: FlutterBinaryMessenger) {
        self.channelName = channelName
        self.channel = FlutterMethodChannel(name: channelName, binaryMessenger: binaryMessenger)
        channel.setMethodCallHandler(handle(_:result:))
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("Invoking: \(call.method), args: \(String(describing: call.arguments))")
        
        let args = call.arguments as! Argument
        let requestId = args["requestId"] as! Int
        let operationType = OperationType(rawValue: call.method)
        
        switch(operationType) {
        case .cancel:
            removeSubscription(requestId: requestId)
        case .subscribe:
            let methodName = args["methodName"] as! String
            let methodType = args["methodType"] as! String
            let arguments = args["arguments"] as! Argument
            
            let type = MethodType(rawValue: methodType)
            switch(type) {
            case .single:
                if registeredSingle[methodName] == nil {
                    result(
                        FlutterError(
                            code: METHOD_NOT_FOUND,
                            message: "Method \(methodName) not registered.",
                            details: nil
                        )
                    )
                    break
                }
                
                let source = registeredSingle[methodName]!
                subscriptions[requestId] = source(arguments)
                    .subscribe { (data: Any) in
                        result(data)
                        self.removeSubscription(requestId: requestId)
                    } onError: { (error: Error) in
                        result(
                            FlutterError(
                                code: OPERATION_ERROR,
                                message: error.localizedDescription,
                                details: nil
                            )
                        )
                        self.removeSubscription(requestId: requestId)
                    }
                break
            case .completable:
                if registeredCompletable[methodName] == nil {
                    result(
                        FlutterError(
                            code: METHOD_NOT_FOUND,
                            message: "Method \(methodName) not registered.",
                            details: nil
                        )
                    )
                    break
                }
                
                let source = registeredCompletable[methodName]!
                subscriptions[requestId] = source(arguments)
                    .subscribe(onCompleted: {
                        result(nil)
                        self.removeSubscription(requestId: requestId)
                    },onError: { (error: Error) in
                        result(
                            FlutterError(
                                code: OPERATION_ERROR,
                                message: error.localizedDescription,
                                details: nil
                            )
                        )
                        self.removeSubscription(requestId: requestId)
                    })
                break
            case .observable:
                guard registeredObservable[methodName] != nil else {
                    result(
                        FlutterError(
                            code: METHOD_NOT_FOUND,
                            message: "Method \(methodName) not registered.",
                            details: nil
                        )
                    )
                    break
                }
                
                let source = registeredObservable[methodName]!
                subscriptions[requestId] = source(arguments)
                    .observeOn(MainScheduler.instance)
                    .do(
                        onCompleted: {
                            result(nil)
                            self.removeSubscription(requestId: requestId)
                        }
                    )
                    .subscribe { (data: Any) in
                        let payload = ObservableCallback(
                            requestId: requestId,
                            value: data
                        )
                        
                        do {
                            self.channel.invokeMethod(
                                "observableCallback",
                                arguments: try payload.toJson()
                            )
                        } catch {
                            print(error.localizedDescription)
                        }
                    } onError: { (error: Error) in
                        result(
                            FlutterError(
                                code: OPERATION_ERROR,
                                message: error.localizedDescription,
                                details: nil
                            )
                        )
                        self.removeSubscription(requestId: requestId)
                    }
                break
            default:
                result(
                    FlutterError(
                        code: INVALID_OPERATION,
                        message: "Invalid method type \(methodType).",
                        details: nil
                    )
                )
            }
            break
        default:
            result(
                FlutterError(
                    code: INVALID_OPERATION,
                    message: "Invalid operation \(call.method).",
                    details: nil
                )
            )
            break
        }
    }
    
    public func registerSingle(_ methodName: String, call:@escaping SingleSource) {
        print("Registered single: \(methodName)")
        registeredSingle[methodName] = call
    }
    
    public func registerCompletable(_ methodName: String, call:@escaping CompletableSource) {
        print("Registered completable: \(methodName)")
        registeredCompletable[methodName] = call
    }
    
    public func registerObservable(_ methodName: String, call:@escaping ObservableSource) {
        print("Registered observable: \(methodName)")
        registeredObservable[methodName] = call
    }
    
    func removeSubscription(requestId: Int) {
        print("Disposing subscription with requestID: \(requestId)")
        subscriptions[requestId]?.dispose()
        subscriptions.removeValue(forKey: requestId)
    }
    
}
